src/open_telemetry_receiver.cc

```diff
@@ -0,0 +1,503 @@
+  label_set_t resource = {0};
+
+  grpc::Status s = unmarshal_resource(rm.resource(), &resource);
+  if (!s.ok()) {
+    return s;
+  }
+
+  for (auto sm : rm.scope_metrics()) {
+    for (auto m : sm.metrics()) {
+      grpc::Status s = dispatch_metric(m, resource, ps);
+      if (!s.ok()) {
+        return s;
+      }
+    }
+  }
+
+  label_set_reset(&resource);
+  return grpc::Status::OK;
+}
+
+/*
+ * OpenTelemetry MetricsService
+ */
+class OTMetricsService : public MetricsService::Service {
+public:
+  grpc::Status Export(grpc::ServerContext *context,
+                      const ExportMetricsServiceRequest *req,
+                      ExportMetricsServiceResponse *resp) {
+    ExportMetricsPartialSuccess *ps = resp->mutable_partial_success();
+
+    for (auto rm : req->resource_metrics()) {
+      grpc::Status s = dispatch_resource_metrics(rm, ps);
+      if (!s.ok()) {
+        ERROR("open_telemetry plugin: dispatch_resource_metrics failed: %s",
+              s.error_message().c_str());
+        return s;
+      }
+    }
+
+    return grpc::Status::OK;
+  }
+};
+
+/*
+ * gRPC server implementation
+ */
+class CollectorServer final {
+public:
+  void Start() {
+#if HAVE_GRPCPP_REFLECTION
+    grpc::reflection::InitProtoReflectionServerBuilderPlugin();
+#endif
+    auto auth = grpc::InsecureServerCredentials();
+
+    grpc::ServerBuilder builder;
+
+    for (auto l : listeners) {
+      grpc::string addr = l.addr + ":" + l.port;
+
+      auto use_ssl = grpc::string("");
+      auto a = auth;
+      if (l.ssl != nullptr) {
+        use_ssl = grpc::string(" (SSL enabled)");
+        a = grpc::SslServerCredentials(*l.ssl);
+      }
+
+      builder.AddListeningPort(addr, a);
+      INFO("open_telemetry plugin: Listening on %s%s", addr.c_str(),
+           use_ssl.c_str());
+    }
+
+    builder.RegisterService(&metrics_service);
+
+    server = builder.BuildAndStart();
+  }
+
+  void Shutdown() { server->Shutdown(); }
+
+private:
+  OTMetricsService metrics_service;
+
+  std::unique_ptr<grpc::Server> server;
+};
+
+static CollectorServer *server = nullptr;
+
+static int receiver_init(void) {
+  if (server) {
+    return 0;
+  }
+
+  server = new CollectorServer();
+  if (!server) {
+    ERROR("open_telemetry plugin: Failed to create server");
+    return -1;
+  }
+
+  server->Start();
+  return 0;
+} /* receiver_init() */
+
+static int receiver_shutdown(void) {
+  if (!server)
+    return 0;
+
+  server->Shutdown();
+
+  delete server;
+  server = nullptr;
+
+  return 0;
+} /* receiver_shutdown() */
+
+static void receiver_install_callbacks(void) {
+  static bool done;
+
+  if (!done) {
+    plugin_register_init("open_telemetry_receiver", receiver_init);
+    plugin_register_shutdown("open_telemetry_receiver", receiver_shutdown);
+  }
+
+  done = true;
+}
+
+// config_get_file is implemented in src/open_telemetry_exporter.cc
+int config_get_file(oconfig_item_t const *ci, grpc::string *out);
+
```

Comment:
Better to have this in a separate header file (e.g. `open_telemetry_common.h`) which is included both by exporter and receiver code.